lsp/lsp.cpp

```diff
@@ -74,25 +89,192 @@ static void sendMessage(const JAST &message) {
+
+  diagnosticJSON.add("message", diagnostic.getMessage());
+
+  return diagnosticJSON;
+}
+
+JAST createDiagnosticMessage() {
+  JAST message = createMessage();
+  message.add("method", "textDocument/publishDiagnostics");
+  return message;
+}
+
+void diagnoseFile(std::string fileUri) {
+  std::string filePath = fileUri.substr(rootUri.length() + 1, std::string::npos);
+  std::unique_ptr<Top> top(new Top);
+
+  auto fileChangesPointer = changedFiles.find(fileUri);
+  if (fileChangesPointer != changedFiles.end()) {
+    Lexer lex(runtime.heap, (*fileChangesPointer).second, filePath.c_str());
+    parse_top(*top, lex);
+  } else {
+    Lexer lex(runtime.heap, filePath.c_str());
+    parse_top(*top, lex);
+  }
+
+  JAST diagnosticsArray(JSON_ARRAY);    
+  for (Diagnostic diagnostic: diagnostics) {
+    diagnosticsArray.children.emplace_back("", createDiagnostic(diagnostic)); // add .add for JSON_OBJECT to JSON_ARRAY
+  }
+  JAST message = createDiagnosticMessage();
+  JAST &params = message.add("params", JSON_OBJECT);
+  params.add("uri", fileUri.c_str());
+  params.children.emplace_back("diagnostics", diagnosticsArray);
+  diagnostics.clear();
+  sendMessage(message);  
+}
+
+void didOpen(JAST receivedMessage) {
+  std::string fileUri = receivedMessage.get("params").get("textDocument").get("uri").value;
+  diagnoseFile(fileUri);
+}
+
+void didChange(JAST receivedMessage) {
+  std::string fileUri = receivedMessage.get("params").get("textDocument").get("uri").value;
+  std::string fileContent = receivedMessage.get("params").get("contentChanges").children.back().second.get("text").value;
+  changedFiles[fileUri] = fileContent;
+  diagnoseFile(fileUri);
+}
+
+void didSave(JAST receivedMessage) {
+  std::string fileUri = receivedMessage.get("params").get("textDocument").get("uri").value;
+  changedFiles.erase(fileUri);
+  diagnoseFile(fileUri);
+}
+
+void didClose(JAST receivedMessage) {
+  std::string fileUri = receivedMessage.get("params").get("textDocument").get("uri").value;
+  changedFiles.erase(fileUri);
+}
+
+void didChangeWatchedFiles(JAST receivedMessage) {
+  JAST files = receivedMessage.get("params").get("changes");
+  for (auto child: files.children) {
+    std::string fileUri = child.second.get("uri").value;
+    changedFiles.erase(fileUri);
+    diagnoseFile(fileUri);
+  }       
+}
+
+void shutdown(JAST receivedMessage) {
+  JAST message = createResponseMessage(receivedMessage);
+  message.add("result", JSON_NULLVAL);
+  isShutDown = true;
+  sendMessage(message);
+}
+
+void serverExit(JAST _) {
+  exit(isShutDown ?0:1);
+}
+
+std::map<std::string, std::function<void(JAST)>> methodToFunction = {
+  {"initialize", initialize},
+  {"initialized", initialized},
+  {"textDocument/didOpen", didOpen},
+  {"textDocument/didChange", didChange},
+  {"textDocument/didSave", didSave},
+  {"textDocument/didClose", didClose},
+  {"workspace/didChangeWatchedFiles", didChangeWatchedFiles},
+  {"shutdown", shutdown},
+  {"exit", serverExit}
+};
```

Comment:
Something about putting function pointers into a table just rubs me the wrong way. I guess it's fine.